src/main/java/muramasa/antimatter/recipe/RecipeMap.java

```diff
@@ -4,25 +4,139 @@
 import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
 import it.unimi.dsi.fastutil.objects.Object2ObjectLinkedOpenHashMap;
 import it.unimi.dsi.fastutil.objects.Object2ObjectMap;
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
+import muramasa.antimatter.Antimatter;
 import muramasa.antimatter.AntimatterAPI;
+import muramasa.antimatter.Ref;
 import muramasa.antimatter.capability.machine.MachineFluidHandler;
 import muramasa.antimatter.capability.machine.MachineItemHandler;
 import muramasa.antimatter.machine.Tier;
 import muramasa.antimatter.registration.IAntimatterObject;
 import muramasa.antimatter.util.Utils;
 import net.minecraft.item.ItemStack;
+import net.minecraft.util.ResourceLocation;
 import net.minecraft.util.text.ITextComponent;
 import net.minecraft.util.text.TranslationTextComponent;
 import net.minecraftforge.fluids.FluidStack;
 
 import javax.annotation.Nullable;
-import java.util.Collection;
-import java.util.Map;
+import java.util.*;
+import java.util.function.BiFunction;
 import java.util.stream.Collectors;
 
 public class RecipeMap<B extends RecipeBuilder> implements IAntimatterObject {
 
+    protected static class RecipeTagMap {
+        //First, match the non-tagged items.
+        private final Object2ObjectMap<RecipeInput, Int2ObjectMap<Recipe>> LOOKUP_TAG;
+        //Tags present, others are ignored.
+        private final Set<ResourceLocation> tagsPresent = new ObjectOpenHashSet<>();
+        public RecipeTagMap() {
+            this.LOOKUP_TAG = new Object2ObjectLinkedOpenHashMap<>();
+        }
+
+        public Recipe find(RecipeInput input) {
```

Comment:
Very nice idea here, it should be fine for now, but would definitely like to seek out a more streamlined solution down the line. But we're here to get the alpha out as soon as possible.